File: 2__cpp_and_python_compare/src/main.cpp
#include <cmath>
#include <cstddef>
#include <iostream>
#include <math.h>
#include <cstring>
#include <chrono>
#include <iomanip>
#include <string>


uint MAX_INT_NUMBERS[] = {7, 4, 6, 3, 8, 4, 7, 4, 1, 2};


double solve_equation(double x){

    return std::pow(x, 2.0) - std::pow(x, 2.0) + std::pow(x, 4.0) - std::pow(x, 5.0) + x + x;
}


int main(int argc, char* argv[]) {

    uint numb_iterations = 0;

    if (argc == 1) return 0; // nothing input

    if (argc != 2) {
        std::cerr << "Error: no input provided\n";
        std::cerr << "Usage: ./program <iterations>\n";
        return 1;
    }

    size_t sz_symb = std::strlen(argv[1]);
 
    // max unsigned int 2 147 483 647
    if (sz_symb > 10){
        std::cerr << "Error: too big number for uint\n";
        return 1;
    }

    uint symbol;
    char symbol_char;
    
    for (size_t i = 0; i < sz_symb; ++i){

        symbol_char = (argv[1])[i];
        
        //is number check
        if (symbol_char < 48 && 57 < symbol_char){
            std::cerr << "Error: input not a number\n";
            return 1;
        } 

        symbol = static_cast<uint>((argv[1])[i]) - 48; // char -> uint , (-48) for this

        uint cur_power = (sz_symb - 1) - i;

        //max in check
        if ((sz_symb == 10) && (MAX_INT_NUMBERS[cur_power] < symbol)){
            std::cerr << "Error: too big number for uint\n";
            return 1;
        }
        
        numb_iterations += symbol * std::pow(10u, cur_power); 
    }

    start_func_iter:

    // solve result and count time
    double result = 0;

    std::chrono::high_resolution_clock::time_point start = std::chrono::high_resolution_clock::now();

    for (uint i = 0; i < numb_iterations; ++i){
        result = solve_equation(result);
    }

    std::chrono::high_resolution_clock::time_point end = std::chrono::high_resolution_clock::now();

    std::chrono::nanoseconds duration = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start);

    std::cout << "Time: " << duration.count() << " nanoseconds\n" <<
     "Func result: " << std::fixed << std::setprecision(16) << result << std::endl;

    char answer;
    std::cout << "Run again?(y/n): ";
    std::cin >> answer;

    answer = std::tolower(static_cast<unsigned char>(answer));

    if (answer == 'y') {
        goto start_func_iter;
    }
    
    return 0;
}

File: 2__cpp_and_python_compare/src/main.py
import time


def calculate(x: float) -> float:
    return x ** 2 - x ** 2 + x * 4 - x * 5 + x + x


def read_iterations():
    user_input = input("Enter number of iterations: ").strip()

    if user_input == "":
        return None

    if not user_input.isdigit():
        return None

    return int(user_input)


def run_calculations(iterations: int) -> None:
    x = 10.0

    start_time = time.perf_counter()

    result = 0.0
    for _ in range(iterations):
        result = calculate(x)

    end_time = time.perf_counter()
    elapsed_time = end_time - start_time

    print(f"Result: {result}")
    print(f"Execution time: {elapsed_time:.10f} seconds")


def main():
    while True:
        iterations = read_iterations()

        if iterations is None:
            print("Input is not a valid non-negative integer. Program finished.")
            break

        run_calculations(iterations)

        answer = input("Run again? (y/n): ").strip().lower()
        if answer != "y":
            print("Program finished.")
            break


if __name__ == "__main__":
    main()